feat(employee listing): Add Employee Listing Page with Dual Views and Advanced Functionalities for Enhanced User Experience

seraGeorge opened the pull request:
## Description

This pull request introduces an Employee Listing Page with dual views (List and Grid) and advanced functionalities to enhance the overall user experience. The key features include:

1. **Dual Views:**
   - Implemented both List and Grid views for flexible user preferences.

2. **Pagination and Infinite Loading:**
   - List view features pagination for organized navigation.
   - Grid view incorporates infinite loading for a seamless user experience.

3. **Search Bar:**
   - Integrated a search bar for efficient searching in both List and Grid views.

4. **Filter by Skills:**
   - Users can filter employees based on their skills for targeted results.

5. **Query Parameters:**
   - Added page, skill, and search values in query parameters to enhance URL tracking and bookmarking.

6. **Sort Functionality:**
   - Implemented sorting functionality for better data organization.

7. **Improved UI:**
   - Made enhancements to the user interface for a more polished and user-friendly design.

8. **Deletion of Employees:**
   - Users can now delete single or multiple employees for effective data management.

9. **Redux Integration:**
   - Configured the Redux store to manage the global state.
   - Implemented core functions for fetching data via Axios using Redux.
   - Set up actions and reducers to manage data flow and state for a more robust architecture.

These additions collectively contribute to a more powerful and user-centric Employee Listing Page, providing a rich set of features and improving the overall usability.

seramg committed:
[add]: header

seramg committed:
Merge branch 'main' of github.com:ashwin8448/HRM_Redux into employee-listing

seramg committed:
[add]: footer

seramg committed:
[add]: button component

seramg committed:
[add]: main heading

seramg committed:
[add]: delete modal and button group wrapper

seramg committed:
[add]: skills chip

seramg committed:
[add]: details component

seramg committed:
[add]: error page and implemented error boundary

seramg committed:
[add]: input error

seramg committed:
[add]: radio button group

seramg committed:
[add]: loader

seramg committed:
[add]: input

seramg committed:
[add]: select and its custom styling

ashwin8448 committed:
[edit]: View employee page

ashwin8448 committed:
Merge branch 'components' of github.com:ashwin8448/HRM_Redux into viewEmployee

seramg committed:
[add]: search bar layout

ashwin8448 committed:
[edit]: View employee page

seramg committed:
[add]: filter select

seramg committed:
[add]: actions bar

seramg committed:
[add]: table head and head button

seramg committed:
[add]: employee table data

seramg committed:
[add]: employee table

Commit:
Merge branch 'employee-listing' of github.com:ashwin8448/HRM_Redux into viewEmployee

Commit:
[edit]:Employee listing page

seraGeorge committed:
[add]: form employee and firebase data interfaces
[add]: actions and actiontypes
[add]: populated employee table using data.json
[change]: helper functions according to the api data

ashwin8448 committed:
Merge branch 'employee-listing' of github.com:ashwin8448/HRM_Redux into viewEmployee

ashwin8448 committed:
[edit]: Employee listing page

ashwin8448 committed:
[Add]:Reducers, actions and cofigured store

seramg committed:
[add]: implementation for fetching employee data

ashwin8448 committed:
[edit]: Made changes to dropdown reducers

seraGeorge committed:
Merge branch 'viewEmployee' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[fix]: employees data fetched interface

seraGeorge committed:
[add]: colors

seraGeorge committed:
Merge branch 'components' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[add]: layout for page title and filters and add new emp btn

seraGeorge committed:
[add]: new design for the button

1. white bg and secondary color for the text
2. added transition on hovering and when active

seraGeorge committed:
Merge branches 'components' and 'employee-listing' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[add]: employee table actions

seraGeorge committed:
[add]: employee table actions for view and delete btn

seraGeorge committed:
[add]: filter count in the btn
[add]: no transition prop to the btn
[remove]: main heading and moved to the header
[add]: button click activity to the employee table actions
[remove]: unnecessary global-padding

seraGeorge committed:
[add]: checkbox column in employee table
[add]: EXTRA_LIGHT_GRAY_COLOR and DARK_PRIMARY_COLOR colors
[adjusted]: z index for overlay
[add]: side filter bar on clicking the filter btn
[add]: icons on table head
[remove]: refresh btn

seraGeorge committed:
[add]: checkbox in the table

seraGeorge committed:
[fix]: list of employees checked

seraGeorge committed:
[add]: delete modal using the checkboxes

seraGeorge committed:
[add]: condition when selected the employee

seraGeorge committed:
[add]: loader design
[add]: interface for data fetched from api
[revamped]: fetching dropdown data
[change]: initialstate for filter data
[add]: loader for each filter select

seraGeorge committed:
[add]: overflow of text

seraGeorge committed:
[add]: wraping of text in delete modal

seraGeorge committed:
[fix]: set skills

seraGeorge committed:
[add]: listingActive for the table and grid

seraGeorge committed:
[add]: active chip

seramg committed:
Merge branch 'components' of github.com:ashwin8448/HRM_Redux into employee-listing

ashwin8448 committed:
[edit]: Active chip and details component

seramg committed:
Merge branch 'components' of github.com:ashwin8448/HRM_Redux into employee-listing

seramg committed:
[add]: grid layout for employees and employee cards

seraGeorge committed:
[add]:type change in actions

seraGeorge committed:
[add]: responsive design for emplyee card list

seraGeorge committed:
[add]: search bar and select button to the card layout

seraGeorge committed:
[add]: pagination for employee listing

seraGeorge committed:
[add]: pagination from the select component in the employee table header

seraGeorge committed:
[add]: table sorting

seraGeorge committed:
[fix]: pagination select component change if page in the pagination bubble change
[add]: firstname as parameter for sorting

seraGeorge committed:
[change]: tooltip hovering action simplified

seraGeorge committed:
[fix]: pagination depth reached issue

[move]: pagination into employeetable

seraGeorge committed:
[remove]:  console in details component
[change]: params to be nullable

seraGeorge committed:
[add]: basic infinite loading

seraGeorge committed:
merged from main

seraGeorge committed:
[add]: for debugging

seraGeorge committed:
[rename]: employees in store to employeesForList

seraGeorge committed:
[fix]: infinite page loading

the grid button is clicked, the page loads as initial loading

seramg committed:
[change]: restructured the data structure

seramg committed:
[move]:employee data fetching in the parent component

seramg committed:
[add]: skills filtering function

seramg committed:
[remove]: unwanted skills filtering setting on mount

seramg committed:
[move]: search and select all in the employee listing
[redirect]: to page 1 when filtering is done
[add]: table overflow with border

seramg committed:
[add]:searching

seramg committed:
[fix]: filtering in employee card list

seramg committed:
[remove]: timeout delay

seraGeorge committed:
[add]:margin for infinite-scroll-loader
[remove]: unwanted print
[remove]: infinite loading state and passed loading from search params
[add]: type in search params

seraGeorge committed:
Merge branch 'employee-listing' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[fix]: infinite loading with filtering and searching

seraGeorge committed:
Merge branch 'employee-listing' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[add]: employee card list

seraGeorge committed:
[remove]: skill loader
and adjusted card list

seraGeorge committed:
[fix]:started fixing employee card list select all

seraGeorge committed:
[add]: necessary log for debugging

seraGeorge committed:
Merge branch 'employee-listing' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[add]: logic for checkbox deleting

seraGeorge committed:
[add]: logic for confirm delete action

seraGeorge committed:
[add]: minor fix to the deete checkbox

seramg committed:
optimised the code

seramg committed:
[remove]: filter select list

seraGeorge committed:
Optimised the code

seramg committed:
[add]: tooltip height to 100px
[remove]:  table loader top to 50%

seramg committed:
[add]: search params updation on deleting modal

seramg committed:
[remove]: unwanted consoles

seraGeorge committed:
[add]: design changes to tooltip

seraGeorge committed:
Merge branch 'employee-listing' of https://github.com/ashwin8448/HRM_Redux into employee-listing

seraGeorge committed:
[rename]: IEmployee to IReceivingEmployee

seraGeorge committed:
[add]: display parameter on search url and
[remove]: unwanted parameters in url

seraGeorge committed:
[add]: sort button

ashwin8448 commented on .github/workflows/deploy.yml:
```diff
@@ -21,8 +21,7 @@ concurrency:
   cancel-in-progress: true
```
Change branch name-"firebase integration"

seraGeorge replied:
Resolved in a7625b4

ashwin8448 commented on .github/workflows/deploy.yml:
```diff
@@ -21,8 +21,7 @@ concurrency:
 env:
-  VITE_DATABASE_URL: ${{ secrets.VITE_DATABASE_URL }}
-  VITE_API_KEY: ${{ secrets.VITE_API_KEY }}
+  VITE_API_ENDPOINT: ${{ secrets.VITE_API_ENDPOINT }}
```
Remove VITE_DATABASE_URL and VITE_API_KEY

seraGeorge replied:
Done

ashwin8448 commented on src/components/Button/button.ts:
```diff
@@ -0,0 +1,68 @@
+  gap: 5px;
+
+  // for the transition
+  backface-visibility: hidden; // hide the element's back face on rotation
```
Is any rotation applied to buttons at any place?

seraGeorge replied:
Resolved in 616328b0ae52390a300a1bc159a1971143c05f9d

ashwin8448 commented on src/components/Button/button.ts:
```diff
@@ -0,0 +1,68 @@
+  // hint for the user agent to prioritize the processing of touch events
+  touch-action: manipulation;
+
+  //TODO: if a different color needs to be added for the icon if there is no text
```
If completed, remove TODO.

seraGeorge replied:
Resolved in c1d862b1e940779d808ab55886138a9ced297f02

ashwin8448 commented on src/components/DeleteModal/constants/constants.ts:
```diff
@@ -1,7 +1,9 @@
+export const CONFIRM_DELETE_TEXT = (employeeIdList: string[]) =>
+  `Are you sure you want to delete the employee${
+    employeeIdList.length > 1 ? "s" : ""
+  } ${employeeIdList.join(", ")} ?`;
```
If possible display employee names also when deleting.

ashwin8448 commented on src/components/Loader/Loader.tsx:
```diff
@@ -0,0 +1,7 @@
+import LoaderWrapper from "./loader.ts";
```
There was an issue with the loader; it was not centered. We can work on it together.

seraGeorge replied:
done

ashwin8448 commented on src/core/config/constants.ts:
```diff
@@ -0,0 +1,13 @@
+// TODO: Delete
+export enum SortDirection {
```
Is this required?

seraGeorge replied:
enum will ensure that sort don't attain any other values other than the ones specified

ashwin8448 commented on src/core/interfaces/interface.ts:
```diff
@@ -0,0 +1,79 @@
+  count: number;
+  loading: boolean;
+}
+export interface IReceivingEmployee {
```
I had made some changes to the interfaces in formPage branch. We can discuss and implement the better one.

seraGeorge replied:
done

ashwin8448 commented on src/core/store/actions.ts:
```diff
@@ -0,0 +1,195 @@
+  type: actionTypes.SET_SKILLS,
+  payload: skills,
+});
+export const resetEmployeesGrid = () => {
```
Why do we need this?

seraGeorge replied:
reseting employees every time we move to the grid

ashwin8448 commented on src/core/store/actionTypes.ts:
```diff
@@ -0,0 +1,13 @@
+export const SET_DATA_EMPLOYEES = "SET_DATA_EMPLOYEES";
+export const SET_EMPLOYEES = "SET_EMPLOYEES";
```
Difference between SET_EMPLOYEES and SET_EMPLOYESS_LIST/SET_EMPLOYEES_GRID?

seraGeorge replied:
SET_EMPLOYEES_LIST // for overriding the current employee list to enable pagination
SET_EMPLOYEES_GRID // for concatenating the current employee list with the new list to enable infinite loading

The reducers deals each action differently

ashwin8448 commented on src/core/store/actions.ts:
```diff
@@ -0,0 +1,195 @@
+  };
+};
+
+export const fetchDropdownData = () => {
```
I guess here the fetch happens sequentially one after the other. We will have to make these fetches simultaneously.

seraGeorge replied:
changed

ashwin8448 commented on src/pages/routes.tsx:
```diff
@@ -1,17 +1,24 @@
+      errorElement: <ErrorPage />,
     },
   ],
   { basename: import.meta.env.DEV ? "/" : "/HRM-App-React/" }
```
Change repo name.

seraGeorge replied:
Resolved in c1d862b1e940779d808ab55886138a9ced297f02

ashwin8448 commented on src/components/Footer/Footer.tsx:
Our header and footer stretches out when we zoom out. We will have to fix it.

seraGeorge replied:
done

ashwin8448 commented on src/core/store/reducers/filterReducer.ts:
Is this file required?

seraGeorge replied:
Resolved in c1d862b1e940779d808ab55886138a9ced297f02

ashwin8448 reviewed (changes requested):
Added few comments. We can work on them together.

seraGeorge committed:
[change]:  branch name in deploy.yml

seraGeorge committed:
[remove]: unwanted css

seraGeorge committed:
[remove]: unwanted todo
[change]: repo name in routes and vite config
[remove]: empty file.

seraGeorge committed:
[change]: tableactions to actions

seraGeorge committed:
[add]: UI Improvements for employee listing page

seraGeorge committed:
[add]: null check for filterselect
[add]: global-width for header
[remove]: sequential loading for dropdown
[rename]: actions->listingactions; actionsbar->filteractions
[add]: handle outside click. if clicked outside dropdown goes off

seraGeorge committed:
[add]: padding to sort order

seraGeorge committed:
[add]: names to the delete modal

seramg committed:
[fix]: infinite loading on reloading

seramg committed:
[remove]: unnecessary code

seramg committed:
[add]:onclick to view page and increase checkbox size

seramg committed:
[remove]: handleActivelisting
